rfcs/text/0000-roadmaps+planning.md

```diff
@@ -0,0 +1,431 @@
+
+***Next Release Milestone 1 Week 3***. Further refinements for Milestone 2 are completely modeled. All Milestone 1 issues that will slip to Milestone 2 MUST be identified.
+
+***Next Release Milestone 2 Week 3***. Further refinements for Milestone 3 are completely modeled. All Milestone 2 issues that will slip to Milestone 3 MUST be identified.
+
+***Next Release Milestone 3 Week 3***.
+All Milestone 3 issues that will slip to the next Release MUST be identified.
+
+#### Tracking Release Progress
+Product Owners for each Objective MUST ensure that their Epics and child issues are kept up-to-date with current status.
+
+Technical Leads MUST drive implementations of prioritized Epics forward and communicate blockers or changes in schedule if they arise.
+
+Over-communication is strongly RECOMMENDED.
+
+***Every two weeks***. The Release Shepherd facilitates a one hour meeting for Product Owners, User Experience, and Technical Leads to review the progress of the Release in ZenHub and address emerging issues.
+
+The Release Shepherd MUST send a status update to the Project Leads on the DCP PM mailing list that captures action items for participants, documents slips in schedule, and/or requests clarification as needed.
+
+As required, the Release Shepherd SHOULD schedule time on the next DCP PM call to address issues pending resolution. The Project Leads MUST never be surprised by the *state of play*.
+
+***At Milestone Completion***. The Project Leads MUST update Oversight on the progress of the Release to ensure appropriate accountability.
+
+#### Modeling progress with ZenHub Pipelines [*dcp* repo]
+
+New items in the *dcp* repo appear in the ZenHub *New* pipeline and MUST be triaged into the *Product Backlog* pipeline, the *Icebox* pipeline, or the *Closed* pipeline by the Release Shepherd or Product Owners.
+
+The ZenHub *Icebox* pipeline SHOULD be used with great restraint in the *dcp* repo. Its explicit purpose is to queue up potential futures from the DCP roadmap such as pending DevSecOps issues. Individual component repo(s)  MAY use the *Icebox* pipeline in a manner of their own choosing.
+
+No issues in the *dcp* repo MUST be assigned to the *Epic* or *Sprint Backlog* pipelines. Individual component repo(s) MAY use these pipelines in a manner of their own choosing.
+
+At the level of the *dcp* repo, a subset of ZenHub pipelines are used to track progress. All Epics or issues in a Release are initially assigned to the *Product Backlog* pipeline. Once the Epic (or any of its children) is started, then its *dcp* issue is moved to the *In Progress* pipeline. When all the children of the Epic are in the *Closed* pipeline, then the *dcp* issue is moved to the *Closed* pipeline.
+
+#### Modeling progress with ZenHub Pipelines [component repo]
```

Comment:
I do not think this RFC should be saying how components should use pipelines in their own repos. This removes flexibility to adapt to local conditions and will not be followed anyway.

Reply:
The intent is to define a minimal and common pipeline workflow so the Product Owners and Technical Leads can understand the progress of an epic when reviewing the top-level ZenHub board:

<img width="715" alt="Screen Shot 2019-08-08 at 10 18 46 AM" src="https://user-images.githubusercontent.com/3644743/62723647-0597ac00-b9c6-11e9-90dd-a11c37591d17.png">

At the very least, we need to have a common understanding on which pipelines signify _In Progress_ and when the child issues are _Done Done_ so the top-level epic can be closed appropriately. 

There's also a new ZenHub feature that allows repo(s) to belong to multiple workspaces with [customized workflows](https://www.zenhub.com/blog/announcing-zenhub-workspaces-personalized-workflows-for-every-team/).

Reply:
Now I am arguing again. Oh well. 

All pipelines in _In Progress_ up to but not including _Closed_ contain tickets that are in progress. That's a simple and easy to understand rule.

- _In Progress_ (tickets whose resolution is currently being implemented)
- _Review_ (tickets whose resolution is currently being reviewed)
- _Dev_ (tickets whose resolution is currently observable in the `dev` deployment of the component)
- _Integration_ (tickets whose resolution is currently observable in the `integration` deployment)
- _Staging_ (tickets whose resolution is currently observable in the `staging` deployment)
- _Prod_ (tickets whose resolution is currently observable in the `integration` deployment)
- _Closed_ (tickets whose resolution as been confirmed by the stake holder)

I think these fall under the definition of "minimal". Which PM wouldn't want to know which tickets are currently observable in the staging deployment, for example?

Azul uses the _Done_ column as a substitute for the above _Dev_ column. We also have a _Merged_  column between `Review` and `Done` but we could do away with that or use a custom workflow as @brianraymor  suggests.

Reply:
Personally, I'd love to have this fine(r) granularity to observe progress. I expect that other teams may have their own preferred workflows based on side conversations. It would help if they would share their opinions here.

Even if specified as a result of _silent consent_, I suspect that not all teams would invest the time in having their Devs keep this up to date. It might require Technical Architecture to document as their **best practice** to encourage compliance. 

The other alternative for this RFC is to indicate that we'll require _Closed_ in component repo(s) to mean it's _Done Done_ and that all other states beyond _Product/Sprint Backlog_ or _Icebox_ mean _In Progress_ when we're reviewing the high-level epics. Some teams will provide us with more detail on progress than others.

Reply:
I'm not sure if this is helpful or not, but Zenhub workspaces allow for teams to have custom 'pipelines' (columns) now. 

So (as an example) let's say currently the DCP workspace has the Azul, Data Portal, Secondary Analysis, Ingest, and Data Store repos connected to it. It has all of the issues linked and has columns A, B, C, D. 
You can make a new workspace "My Fave Issues" and link only say the two repos your team works on and tracks tickets in, let's say Azul and Data Portal. On your workspace, you can have columns A, C, E, F, G, H, I and track ticket progression through those columns however you please. It has no effect on the DCP workspace and enables you to work however you'd like. 

The catch is that the status of the tickets in the repos Azul and Data Portal in My Fave Issues won't be reflected in the DCP workspace. So someone would have to move those tickets around should we want to reflect status at that level in both places.

The reason Zenhub has it implemented this way is if let's say a design team and an implementation team are both working on Feature X, but want to reflect different statuses on their progress on Feature X in their workspace view of the board, the design team can say 'we're `in progress` on designing this' and the implementation team could say 'we have this in the `backlog`' with the same exact issue. No need to duplicate one for each team. It's actually kinda cool!!! 

Here's documentation on this new feature (it was only added in May)
https://help.zenhub.com/support/solutions/articles/43000495236-workspace-use-cases

And you can see a [test board](https://app.zenhub.com/workspaces/secondary-analysis-5d4c8e49bb04cc6a337a120d/board?repos=85364092,99818644,110702686,138780916,110603502,108927219,112664027,118922733,166075280) I made here to try it out, linking all the pipelines related repos, where one of them is also on the DCP workspace, and in this workspace I have different columns.

Reply:
I also referenced it earlier in the [thread](https://github.com/HumanCellAtlas/dcp-community/pull/92#discussion_r312154399), but thanks for the additional information and test board which clarifies.

Reply:
When you look at an issue you can see on the right which pipeline it's in per workspace, and you can move it in both workspaces from the issue view. 
![image](https://user-images.githubusercontent.com/9449764/62738484-6ba12500-ba00-11e9-807f-babe35a3ec06.png)

Reply:
If an issue in secondary-analysis is part of a DCP workspace epic, looking at the epic status view that Brian screenshotted above would reflect the status of the ticket in the DCP workspace.

Reply:
If teams are interested in having their own choices of columns, we could make it so the DCP workspace has columns that make more sense for the project wide management, maybe.. New, Icebox, Quarterly Release Plan, In Progress, In Prod, Closed?

Reply:
Thank you for the analysis, @kbergin. If this RFC describes the a workspace predominantly for the PMs, and the PMs don't see any value in knowing at a glance what tickets are in what deployment—it would puzzle me if they didn't, but I'd accept it, of course—I will rest my case.

Reply:
For my part, I am not using the Merged pipeline because the amount of ticket management is too high for me. I use the Done pipeline for accounting purposes to know what was achieved in a sprint before I close the tickets at the end. I also have to deal with data operations tickets which this RFC does not cover but does complicate my board management.

Others may think this is a wrong use of the pipelines but it works for us. My point is that, to [quote](https://internals.rust-lang.org/t/governance-working-group-announcement/9637) the Rust Governance Working Group

"""
One of the Rust project’s current strengths is that it allows teams to find their own workflows. Processes are rarely a “one size fits all” sort of thing.
"""

and 

"""
We like to see this working group working with the teams it is trying to help, studying what they do, proposing experiments, and synthesizing best practices and facilitate cross-team communication and exchange.
"""

So I have no objection with seeking alignment over time but this shouldn't be through an RFC until common team practice has emerged through communication and exchange.

That said, I understand the need for information to come up through the boards. But I think that is already achieved through in progress and closed tickets. The finer details always need interpretation anyway and that's what the humans are for.

Reply:
@hannes-ucsc I can see where you've coming from with those pipelines but at the moment I'm finding even Merged too much overhead, it being easier to track progress through flow of release tags through the system. I may well play with your approach but mandating it through an RFC is not the way to go in my opinion.

Reply:
I've performed some initial experiments in multiple workspaces after reviewing - https://help.zenhub.com/support/solutions/articles/43000484539-issue-placement-when-adding-repos-to-a-workspace- . There is no ZenHub API to automate the process of copying a repo to a second workspace; however, the UI is relatively simple. I was able to add the matrix-service to a second workspace while maintaining pipelines and order from the DCP workspace. There are some minor UI bugs with easy workarounds that I can document.